File: quattro2/mac/Common/RolandWirelessConnect/RWCAudioOutputStream.m
//
//  RWCAudioOutputStream.m
//
//  Copyright 2012 Roland Corporation. All rights reserved.
//

#import "RWCConnection.h"

#define OUTPUT_BUF_SIZE (8192)

@interface RWCAudioOutputStream () {

    RWCConnection *_connection;
    RWCNetServer *_server;

    NSInteger _length;
    NSInteger _offset;
    u_int8_t  _buffer[OUTPUT_BUF_SIZE];
}
@end

@implementation RWCAudioOutputStream

@synthesize delegate = _delegate;

@dynamic status;
@dynamic numberOfChannels;
@dynamic currentTime;
@dynamic volume;
@dynamic speed;

- (id)initWithConnection:(RWCConnection *)connection
{
    if (self = [super init]) {
        _connection = connection;
        _server = [[RWCNetServer alloc] initWithPortNo:0];
        _server.delegate = self;
    }
    return self;
}

- (void)dealloc
{
    _connection = nil;
    _server.delegate = nil;
    [_server release];
    [super dealloc];
}

- (void)startServer
{
    [_server startServer];
}

- (void)stopServer
{
    [_server stopServer];
}

- (void)ASBD:(AudioStreamBasicDescription *)format
{
    UInt32 channel = kRWCNumOfChannels;
    
    format->mSampleRate        = kRWCSampleRate;
    format->mFormatID          = kAudioFormatLinearPCM;
    format->mFormatFlags       = kLinearPCMFormatFlagIsSignedInteger
    | kLinearPCMFormatFlagIsPacked;
    format->mChannelsPerFrame  = channel;
    format->mBytesPerPacket    = sizeof(int16_t) * channel;
    format->mBytesPerFrame     = sizeof(int16_t) * channel;
    format->mBitsPerChannel    = 8 * sizeof(int16_t);
    format->mFramesPerPacket   = 1;
    format->mReserved          = 0;
}

- (void)start:(NSString *)deviceUID
{
    u_int8_t buf[16];

    RNP_HEADER *header = (RNP_HEADER*)buf;
    header->type    = RNP_TYPE_AUDIO_CONTROL;
    header->subtype = RNP_SUBTYPE_OUTPUT_START;
    header->size[0] = 0;
    header->size[1] = 2;

    RNP_AUDIO_CONTROL *data = (RNP_AUDIO_CONTROL*)(buf + sizeof(RNP_HEADER));
    data->value[0] = (unsigned char)(_server.portNo >> 8);
    data->value[1] = (unsigned char)(_server.portNo);

    [_connection writeToServer:buf];
}

- (void)pause
{
    u_int8_t buf[16];

    RNP_HEADER *header = (RNP_HEADER*)buf;
    header->type    = RNP_TYPE_AUDIO_CONTROL;
    header->subtype = RNP_SUBTYPE_OUTPUT_PAUSE;
    header->size[0] = 0;
    header->size[1] = 0;

    [_connection writeToServer:buf];
}

- (void)stop
{
    u_int8_t buf[16];

    RNP_HEADER *header = (RNP_HEADER*)buf;
    header->type    = RNP_TYPE_AUDIO_CONTROL;
    header->subtype = RNP_SUBTYPE_OUTPUT_STOP;
    header->size[0] = 0;
    header->size[1] = 0;

    [_connection writeToServer:buf];

    [_server stopConnection];
}

- (int)status
{
    if (_server.outputStream) {
        return _connection.audioStatus->output.status;
    }
    return kAudioOutputStatusStop;
}

- (NSUInteger)numberOfChannels
{
    return kRWCNumOfChannels;
}

- (NSTimeInterval)currentTime
{
    unsigned long currentDeviceFrame =
        (_connection.audioStatus->output.frame[0] << 24) |
        (_connection.audioStatus->output.frame[1] << 16) |
        (_connection.audioStatus->output.frame[2] <<  8) |
        (_connection.audioStatus->output.frame[3] <<  0);

    return currentDeviceFrame / kRWCSampleRate;
}

- (float)volume
{
    return ((float)_connection.audioStatus->output.volume / 127.0);
}

- (void)setVolume:(float)gain
{
    if (gain < 0 || gain > 1.0)
        return;

    u_int8_t buf[16];

    RNP_HEADER *header = (RNP_HEADER*)buf;
    header->type    = RNP_TYPE_AUDIO_CONTROL;
    header->subtype = RNP_SUBTYPE_OUTPUT_VOLUME;
    header->size[0] = 0;
    header->size[1] = 1;

    RNP_AUDIO_CONTROL *data = (RNP_AUDIO_CONTROL*)(buf + sizeof(RNP_HEADER));
    data->value[0] = (unsigned char)(gain * 127);

    [_connection writeToServer:buf];
}

- (float)speed
{
	return 1.0f;
}

- (void)setSpeed:(float)rate
{
	/* not supported */
}

- (float)pitch
{
	return 0.0f;
}

- (void)setPitch:(float)cent
{
	/* not supported */
}

- (float)peakPowerForChannel:(NSUInteger)channelNumber
{
    if (channelNumber >= kRWCNumOfChannels)
        return 0;

    return [_connection outputLevel:(int)channelNumber];
}

- (void)streamEventOpenCompleted:(RWCNetServer *)session
{
    _offset = _length = 0;
}

- (void)streamHasSpaceAvailable:(RWCNetServer *)session stream:(NSOutputStream *)ostream
{
    if (_length == 0) {
        _offset = 0;
        _length = [self.delegate outputStream:_buffer maxLength:sizeof(_buffer)];
    }

    NSInteger bytes = [ostream write:_buffer + _offset maxLength:_length];
    if (bytes <= 0) {
        [_server stopConnection];
        _length = 0;
        return;
    }

    _length -= bytes;
    _offset += bytes;
}

- (void)streamHasBytesAvailable:(RWCNetServer *)session stream:(NSInputStream *)istream
{
    u_int8_t dummy;
    [istream read:&dummy maxLength:sizeof(dummy)];
}

- (void)streamErrorOccurred:(RWCNetServer *)session { }
- (void)streamEndEncountered:(RWCNetServer *)session { }

@end

File: quattro2/win10/RolandWirelessConnect/RWCAudioOutputStream.cpp
//
// @(#)RWCAudioOutputStream.cpp
//
// Copyright 2013 Roland Corporation. All rights reserved.
//

#include "RWCConnection.h"

RWCAudioOutputStream::RWCAudioOutputStream(RWCConnection* conn)
{
	mp_connection = conn;
	mp_delegate = 0;

	_length = _offset = 0;

	memset(&m_wfe, 0, sizeof(m_wfe));
	m_wfe.wFormatTag      = WAVE_FORMAT_PCM;
	m_wfe.nSamplesPerSec  = (DWORD)kRWCSampleRate;
	m_wfe.nChannels       = kRWCNumOfChannels;
	m_wfe.wBitsPerSample  = (kRWCBytesPerFrame / kRWCNumOfChannels) * 8;
	m_wfe.nBlockAlign     = kRWCBytesPerFrame;
	m_wfe.nAvgBytesPerSec = kRWCBytesPerFrame * m_wfe.nSamplesPerSec;
}

void RWCAudioOutputStream::start(LPCTSTR)
{
	acceptConnection();

	BYTE buf[16];

	RNP_HEADER *header = (RNP_HEADER*)buf;
	header->type    = RNP_TYPE_AUDIO_CONTROL;
	header->subtype = RNP_SUBTYPE_OUTPUT_START;
	header->size[0] = 0;
	header->size[1] = 2;

	RNP_AUDIO_CONTROL *data = (RNP_AUDIO_CONTROL*)(buf + sizeof(RNP_HEADER));
	data->value[0] = (unsigned char)(portNo() >> 8);
	data->value[1] = (unsigned char)(portNo());

	mp_connection->writeToServer(buf);
}

void RWCAudioOutputStream::pause()
{
    BYTE buf[16];

    RNP_HEADER *header = (RNP_HEADER*)buf;
    header->type    = RNP_TYPE_AUDIO_CONTROL;
    header->subtype = RNP_SUBTYPE_OUTPUT_PAUSE;
    header->size[0] = 0;
    header->size[1] = 0;

	mp_connection->writeToServer(buf);
}

void RWCAudioOutputStream::stop()
{
    BYTE buf[16];

    RNP_HEADER *header = (RNP_HEADER*)buf;
    header->type    = RNP_TYPE_AUDIO_CONTROL;
    header->subtype = RNP_SUBTYPE_OUTPUT_STOP;
    header->size[0] = 0;
    header->size[1] = 0;

	mp_connection->writeToServer(buf);

	stopConnection();
}

int RWCAudioOutputStream::status() const
{
	if (connecting()) {
		return mp_connection->_audioStatus.output.status;
	}
	return RNP_AUDIO_STATUS_STOP;
}

int RWCAudioOutputStream::numberOfChannels() const
{
	return kRWCNumOfChannels;
}

float RWCAudioOutputStream::currentTime() const
{
	DWORD currentDeviceFrame =
		(mp_connection->_audioStatus.output.frame[0] << 24) |
		(mp_connection->_audioStatus.output.frame[1] << 16) |
		(mp_connection->_audioStatus.output.frame[2] <<  8) |
		(mp_connection->_audioStatus.output.frame[3] <<  0);

	return (float)(currentDeviceFrame / kRWCSampleRate);
}

float RWCAudioOutputStream::volume() const
{
	return (float)(mp_connection->_audioStatus.output.volume / 127.0);
}

void RWCAudioOutputStream::volume(float gain)
{
	if (gain < 0 || gain > 1.0)
		return;

	BYTE buf[16];

	RNP_HEADER *header = (RNP_HEADER*)buf;
	header->type    = RNP_TYPE_AUDIO_CONTROL;
	header->subtype = RNP_SUBTYPE_OUTPUT_VOLUME;
	header->size[0] = 0;
	header->size[1] = 1;

	RNP_AUDIO_CONTROL *data = (RNP_AUDIO_CONTROL*)(buf + sizeof(RNP_HEADER));
	data->value[0] = (unsigned char)(gain * 127);

	mp_connection->writeToServer(buf);
}

float RWCAudioOutputStream::peakPowerForChannel(int channelNumber) const
{
	if (channelNumber < 0 || channelNumber >= kRWCNumOfChannels)
		return 0;

	return mp_connection->outputLevel(channelNumber);
}

void RWCAudioOutputStream::streamOpenCompleted()
{
	_length = _offset = 0;
}

bool RWCAudioOutputStream::streamRunLoop(Socket* socket)
{
	if (mp_delegate == 0)
		return false;

	if (_length == 0) {
		_offset = 0;
		_length = mp_delegate->outputStream(_buffer, sizeof(_buffer));
	}

	int len = socket->send(_buffer + _offset, _length);
	if (len <= 0) {
		return false;
	}

	_length -= len;
	_offset += len;

	return true;
}
